apps/builder/app/builder/shared/assets/use-assets.tsx

```diff
@@ -31,68 +33,108 @@ const setAsset = (asset: Asset) => {
   });
 };
 
-type FileData = {
-  assetId: string;
-  type: AssetType;
-  file: File;
-  objectURL: string;
-};
+type UploadingFileData = Simplify<
+  {
+    // common props
+    assetId: string;
+    type: AssetType;
+    objectURL: string;
+  } & (
+    | {
+        source: "file";
+        file: File;
+      }
+    | {
+        source: "url";
+        url: URL;
+      }
+  )
+>;
```

Comment:
not simple 😉

Reply:
After Simplify on hover it's simple type ;-)

<img width="491" alt="image" src="https://github.com/webstudio-is/webstudio/assets/5077042/d3f66002-4bd2-4c44-977f-2fac5dc6a5ab">